pkg/virtualnode/resource_manager.go

```diff
@@ -0,0 +1,276 @@
+
+// findVirtualServiceDependencies find the VirtualService dependencies for this virtualNode.
+func (m *defaultResourceManager) findVirtualServiceDependencies(ctx context.Context, vn *appmesh.VirtualNode) (map[appmesh.VirtualServiceReference]*appmesh.VirtualService, error) {
+	vsByRef := make(map[appmesh.VirtualServiceReference]*appmesh.VirtualService, len(vn.Spec.Backends))
+	for _, backend := range vn.Spec.Backends {
+		vsRef := backend.VirtualService.VirtualServiceRef
+		vs, err := m.virtualServiceRefResolver.Resolve(ctx, vn, vsRef)
+		if err != nil {
+			return nil, errors.Wrapf(err, "failed to resolve virtualServiceRef")
+		}
+		vsByRef[vsRef] = vs
+	}
+	return vsByRef, nil
+}
+
+// validateVirtualServiceDependencies validates the VirtualService dependencies for this virtualNode.
+// AppMesh API allows to create a virtualNode with virtualService backend without the virtualService presents first.
+// we will not validate whether virtualService is active or not, to allows circular dependency between virtualNode and virtualService.
+func (m *defaultResourceManager) validateVirtualServiceDependencies(ctx context.Context, ms *appmesh.Mesh, vsByRef map[appmesh.VirtualServiceReference]*appmesh.VirtualService) error {
+	for _, vs := range vsByRef {
+		if vs.Spec.MeshRef == nil || !mesh.IsMeshReferenced(ms, *vs.Spec.MeshRef) {
+			return errors.Errorf("virtualService %v didn't belong to mesh %v", k8s.NamespacedName(vs), k8s.NamespacedName(ms))
+		}
+	}
+	return nil
+}
+
+func (m *defaultResourceManager) findSDKVirtualNode(ctx context.Context, ms *appmesh.Mesh, vn *appmesh.VirtualNode) (*appmeshsdk.VirtualNodeData, error) {
+	resp, err := m.appMeshSDK.DescribeVirtualNodeWithContext(ctx, &appmeshsdk.DescribeVirtualNodeInput{
+		MeshName:        ms.Spec.AWSName,
+		MeshOwner:       ms.Spec.MeshOwner,
+		VirtualNodeName: vn.Spec.AWSName,
+	})
+	if err != nil {
+		if awsErr, ok := err.(awserr.Error); ok && awsErr.Code() == "NotFoundException" {
+			return nil, nil
+		}
+		return nil, err
+	}
+
+	return resp.VirtualNode, nil
+}
+
+func (m *defaultResourceManager) createSDKVirtualNode(ctx context.Context, ms *appmesh.Mesh, vn *appmesh.VirtualNode, vsByRef map[appmesh.VirtualServiceReference]*appmesh.VirtualService) (*appmeshsdk.VirtualNodeData, error) {
+	sdkVNSpec, err := m.buildSDKVirtualNodeSpec(ctx, vn, vsByRef)
+	if err != nil {
+		return nil, err
+	}
+	resp, err := m.appMeshSDK.CreateVirtualNodeWithContext(ctx, &appmeshsdk.CreateVirtualNodeInput{
+		MeshName:        ms.Spec.AWSName,
+		MeshOwner:       ms.Spec.MeshOwner,
+		Spec:            sdkVNSpec,
+		Tags:            nil,
+		VirtualNodeName: vn.Spec.AWSName,
+	})
+	if err != nil {
+		return nil, err
+	}
+	return resp.VirtualNode, nil
+}
+
+func (m *defaultResourceManager) updateSDKVirtualNode(ctx context.Context, sdkVN *appmeshsdk.VirtualNodeData, ms *appmesh.Mesh, vn *appmesh.VirtualNode, vsByRef map[appmesh.VirtualServiceReference]*appmesh.VirtualService) (*appmeshsdk.VirtualNodeData, error) {
```

Comment:
This is only called in the `Reconcile`, but it might be worth a comment that `IgnoreLeftHandUnset` is used when checking for equality. If the `desiredSDKVNSpec` is missing some settings, what does that mean?

Reply:
if an optional field is not set, AWS will provide default settings(and will be in actualSDKVNSpec). And compare will fail always.
This allows to equality to still be true if specific optional settings is omitted, which allows AWS sever-side to provide default settings.

Reply:
Thanks, that explains it! Might be worth adding as a go doc comment since this is not obvious:
```go
// updateSDKVirtualNode updates the SDK Virtual Node. If an optional field is not set, AWS will provide default 
// settings that will be in actualSDKVNSpec. We use IgnoreLeftHandUnset when doing the equality check here to
// allow AWS sever-side to provide default settings.

```

Reply:
thanks, added comments :D